File: 0_4_IP/Normalize_array/UsersCode.m
function UsersCode(up)

	Methods

	% Enter user code here
	% Find max and min values in array data
	valMax = -0.0000001;
	valMin = 9999999999999.9;

	for index = TopDepth():BottomDepth()
		arrayValues = Array_InCrv_Chunk(index, 1, 1, Array_InCrv_MaxX(), Array_InCrv_MaxY());
		valMin = min(arrayValues(:));
		valMax = max(arrayValues(:));		
	end

	% Calculate the gain and shift to normalize
	gain = 1.0 / (valMax - valMin);
	shift = -gain * valMin;

	% Output new array with normalized values
	for index = TopDepth():BottomDepth()
		arrayValues = Array_InCrv_Chunk(index, 1, 1, Array_InCrv_MaxX(), Array_InCrv_MaxY()) * gain + shift;
		Save_Array_OutCrv_Chunk(index, 1, 1, arrayValues, size(arrayValues, 1), size(arrayValues, 2));
	end

	Save_OutCrv_Comments('Updated by Normalize Array User Prog');
end

File: 0_4_IP/Normalize_array/UsersCode.pas
unit UsersCode;

interface

uses Functions, Math;

Procedure UserCode; stdcall;

implementation

{User Code}
Procedure UserCode; stdcall;
Var
   index, Ix, IY : Integer;
   gain, shift, R, valmin, valmax : Single;
begin
   {enter user code here}
   {find max and min values in array data}
      ValMax := -0.0000001;
      ValMin := 9999999999999.9;        
      for index := Topdepth to BottomDepth do begin
         for IX := 1 to Array_InCrv_MaxX do begin
            for IY := 1 to Array_InCrv_MaxY do begin
               R := Array_InCrv(Index,Ix,Iy);
               if (R < ValMin) and (R <> -999.0) then
                  ValMin := R;
               if (R > ValMax) and (R <> -999.0) then
                  ValMax := R;
            end;
         end; 
      end;
   {calculate the gain and shift to normalize}
      gain := 1.0 / (Valmax - Valmin);
      shift := - gain * Valmin; 
   {output new array with normalized values}
      for index := Topdepth to BottomDepth do begin
         for IX := 1 to Array_InCrv_MaxX do begin
            for IY := 1 to Array_InCrv_MaxY do begin
               R := Array_InCrv(Index,Ix,Iy) * Gain + shift;
               Save_array_OutCrv(Index,Ix,Iy,R);
            end;
         end;
      end;
   Save_OutCrv_Comments('Updated by Normalize Array User Prog');
end;

end.
